### ISSUE TITLE
what filesystem for chunkserver (XFS or ZFS?)

### ISSUE THREAD

**Issue by SirNolaan** [lizardfs/lizardfs#797]
Hello,

What file system do you recommend for chunkserver for Linux Debian 9?
XFS or ZFS?
What mount options for ZFS are the same as for XFS?

**Comment by njhurst** [comment]
ext4 - simple, widely used, reliable and fast.

ZFS would be an odd choice, as you'd waste all that consistency stuff.

**Comment by Zorlin** [comment]
I'm curious as to why to avoid XFS, as both MooseFS and LizardFS teams seem
to recommend it

**Comment by cloudweavers** [comment]
We have 250+ installations with LizardFS, and have used XFS for a long time before switching to ZFS.
XFS is a rock, very stable, can recover from quite damaged disks; performance is on par or better than ext4 with no need for tuning.
ZFS is even more resilient, online consistency checks are capable of finding controller or connector-induced errors, with some tuning can be as fast or faster than XFS. We use it with a single pool per disk, no RAID or RAIDz. Compression is quite efficient, and next version will have encryption as well (we are testing it right now, is already out in beta).

**Comment by onlyjob** [comment]
I specifically tested performance of _ext4_ versus _xfs_ on identical disks on the same chunkserver. _ext4_ is faster on metadata operations and it uses available space better than _xfs_. Anecdotal example was when I've taken 80% utilised _ext4_ HDD from chunkserver and tried to `rsync` it to identical _xfs_ HDD, running out of space on _xfs_ before all the data were copied...

Basically _ext4_ performs better with many files therefore it is preferable when I/O latency matters. Try it yourself on the simplest test: make many files/folders on _xfs_ and _ext4_ then see how long will it take to walk the file system with `time tree .`.

Also with _ext4_ on rotational disks one can use external journal on faster devices (e.g. SSD) which can not be done with _xfs_.

Native _ext4_ encryption is pretty cool, dramatically outperforming xfs-on-LUKS.

Finally, there is [reliability concern](https://raid6.com.au/posts/fs_xfs/) about XFS but that shouldn't stop XFS fans if their chunkservers are on redundant power.  ;)

As for why they (still?) recommend _xfs_ I have no answer. Maybe they follow old enterprise recommendations? Maybe they've never concluded benchmarks? 

The only situation when I would consider to use _xfs_ is for several large files (hundreds of gibibytes) per partition...

As for ZFS, I was always reluctant to use FUSE file systems for chunkserver's data and I prefer to avoid building kernel modules to access chunkservers' partitions.
Not sure if ZFS really makes the difference for chunkservers...
IMHO chunkservers should support native compression: #366.

**Comment by cloudweavers** [comment]
Unless you use --inplace, rsync will always create additional files before "committing" them - so it may explain the difference. We found less than 2% difference in space availability between ext4 and xfs (removing the space for the journal from the equation, obviously).
ext4 and xfs are mostly equal in benchmarks: https://www.phoronix.com/scan.php?page=article&item=linux414-fs-compare&num=2
As for reliability, the link mentions "We've evaluated XFS at the end of 2012". We are in 2019, and XFS improved quite a bit since 2012 (as ext4 did, of course). So I would suggest to reassess this using more recent sources.
As for ZFS, we use ZFSonLinux, which has a native kernel module. As for making a difference, when ZFS warns you that a SCSI controller with bad memory is making bad writes, and natively intercepts and corrects that for you, I can tell you that this makes quite a difference in reliability.

**Comment by onlyjob** [comment]
That was not an `rsync` issue; _xfs_ really does consume significantly more space than _ext4_ by default (and that seems to depend on number of files in file system). 
Maybe that could be tweaked on file system creation but that's one of the reasons why it make sense to prefer _ext4_ especially with many disks.

> and XFS improved quite a bit since 2012

So they've said in 2012...
Back then _xfs_ was considered "stable, enterprise file system" yet _xfs_ developers couldn't care less about reliability in situations with sudden loss of power. :(
I recall reading about similar issues multiple times.

Basically _xfs_ survives power loss without structural damage yet losing far to many changes before crash. Personally I've found it difficult to trust file system developed without consideration to safety for power loss.

Anyway current situation is that _xfs_ is slower than _ext4_ with metadata operations, like deleting files, etc. Even without reliability concerns, less space efficiency and slower metadata operations comparing to _ext4_ make _xfs_ rather unattractive choice for LizardFS chunksevers.

**Comment by borkd** [comment]
@onlyjob - [Native ZFS for Linux](https://zfsonlinux.org/) has been around for close to a decade and is a pretty solid choice for chunkservers, masters, and metaloggers. Were you using it, taking a snapshot prior to upgrade you described in #746 would save you bunch of time and prevent data loss. And, yes, you get compression for free as well.

**Comment by borkd** [comment]
@onlyjob - As for XFS, both performance and post-failure behavior has improved over the years. On chunkservers, a periodic online defrag will help with sequential reads. Self-Describing Metadata makes it possible to detect corruption in metadata after unexpected power loss, which is not a bad thing either. Having a strong personal opinion is one thing, but there is no point in giving advice based on outdated / no longer relevant information.

**Comment by onlyjob** [comment]
How many files have you created?
When only 1% of file system is used the will be hardly any difference in amount of available space...

I've used ~4 TiB of real data from chunkserver's HDD and I've copied it to newly created _xfs_ on identical disk: _xfs_ partition filled when about 80% of all files from _ext4_ partition were copied. There were between 4 and 5 million files on disk, all chunk files no bigger than 65M.

**Comment by cloudweavers** [comment]
I just made a test on a 110GB virtual disk, attached to a Centos 7.4 VM. Using 64MB files with random data, I created 1717 files before stopping with out of space error - 109GB. Certainly not 20%.
Since you talk of 4M files on a 4TB disk, it means that on average you are storing 1Mb per chunk, so you are probably reaching some metadata limit. I don't know what OS version you tested on, so I can't say, but probably it's some less-than-optimal default for 4M files.
I had a mail server with XFS on 4TB storage, with 8M+ files, and reached 92% once (when the purge process was not working properly), so I would say that this is not XFS specific.

**Comment by borkd** [comment]
With small files and close to a full disk, the effects of speculative preallocation were likely seen. There are ways to tune around it, but operating filesystems at close to full condition rarely makes sense from performance point of view, especially when one is "hydrating" sparse files (as it would be common during VM lifecycle). See [XFS FAQ 45-49](http://xfs.org/index.php/XFS_FAQ#Q:_Why_do_files_on_XFS_use_more_data_blocks_than_expected.3F).

**Comment by onlyjob** [comment]
> what OS version you tested on

Debian "Stretch" on amd64 (a.k.a. x86_64).

>  I would say that this is not XFS specific.

Denial detected. :) How could it be not XFS specific if issue manifests on XFS only?
I say it __is__ XFS specific because it have something to do with how allocation works on XFS.

**Comment by hradec** [comment]
I have a home "NAS" like server where I use LFS with ext4 disks. 
The last disk I added, I decided to try ZFS to see... WHUAU!! LFS performance on ZFS is WAY faster than on ext4. I started to re-format all my disks to ZFS, and LFS reading/writing has improved a lot... about 30%!

I'm actually formatting EACH disk as its own ZFS pool! So I'm not doing ZRAID... just one disk pools! 

I also setup one chunkserver per disk on proxmox containers, basically using LFS as a software RAID. 

There's one problem with ZFS one disk pool... Since the pool have no redundancy, if you get bad sectors on the disk, and metadata is written to that bad sector, you can endup with a corrupted drive... and it's not as simple as running "e2fsck" to recover the data! If there's a lot of Metadata corruption, ZFS may prevent the pool from being mounted, which can be scaring! (that happened to me once)

Off course, using LFS with at least ec(N,2) or a 3 goal, loosing one whole disk this way is no big deal, at least for me. 

Also, apart from that one disk, I had a couple of other ZFS single disks pools which showed bad sectors, and ZFS was able to recuperated itself without any problems. A simple scrub fixed the data, moving it out of the bad sectors. I'm still using those disks today as we speak! 

ZFS single disk pool is much faster than ext4... WAY FASTER, and when problem occurs, fixing it is actually simpler since you can scrub it mounted! I really hate the fact that you need to stop the disk (unmount) to run e2fsck!

Regarding XFS, I stay away from it... in my experience, XFS is faster than ext4 for LFS, yes. BUT, when problem occurs, XFS fixing "needs attention"... you can't just run it with "-y" to automatic fix things, like e2fsck. With ext4, "e2fsck -y" is pretty OK. I never lost anything by blindly running "e2fsck -y". 

With XFS, xfs_repair has no "-y" option, and it takes a loooooooooong time to check... so you need to there to answer questions, or else the repair will halt waiting for input! 

ZFS is beautifull in this sense... just run scrub and voila! 

Hoo... don't forget to turn compression ON for ZFS... it not only saves space, but compression actually makes ZFS faster!! Less wait for I/O!! 

And don't turn de-duplication on, unless you have TONS of memory on you machine! (I could never use it, so I can't say if it makes it faster or slows it down!)

my 2 cents....

**Comment by szycha76** [comment]
For already-compressed or encrypted data - I'll consider to use XFS.
For any case, especially for non-compressed, unencrypted data - I'd use ZFS.
For experimental purposes with non-critical data I'd use Ext4, especially after I read this comment: https://lwn.net/Articles/704201/

**Comment by onlyjob** [comment]
What bothers you, specifically, in Theodore's interview?

**Comment by szycha76** [comment]
It was the long comment I was referring to. If you cannot afford ZFS (less than 512 MB RAM, or less than 64-bit O/S) - use XFS. It is way better in the long term. I'd avoid ext4 unless I had a very good reason to try it.
Also I'd suggest giving lz4 compression a shot: `zfs set compression=lz4`, since it will help you save both time and space (at the expense of a few CPU cycles)

**Comment by onlyjob** [comment]
@szycha76, your opinions and preferences are not too interesting, unless they make sense. You are not telling us what is your problem with _ext4_...

**Comment by pbeza** [comment]
Just for sake of completeness: [LizardFS Documentation (page 17)](https://media.readthedocs.org/pdf/lizardfs-docs/latest/lizardfs-docs.pdf) actually recommends [XFS](https://en.wikipedia.org/wiki/XFS) or [ZFS](https://en.wikipedia.org/wiki/ZFS) file systems on servers. Unfortunately there is no rationale mentioned there.

**Comment by szycha76** [comment]
A word of warning I owe you.
Never use non-mirrored ZFS on unreliable medium like usb-attached disk. I developed a ZFS corruption that hit several chunksNN directories on 3 drives. Despite of ec(4,2) parity I lost some data that way.

So now I'm in the process of migration to XFS, since LizardFS utilizes sparse files on XFS pretty well -- see example below:
```
8.0K 65K chunk_ec_1_of_4_2_000000000000002E_00000001.liz
 12K 65K chunk_ec_1_of_4_2_0000000000000198_00000001.liz
 16K 65K chunk_ec_1_of_4_2_0000000000000055_00000001.liz
 28K 65K chunk_ec_1_of_4_2_000000000000010A_00000001.liz
 52K 65K chunk_ec_1_of_4_2_0000000000000121_00000001.liz
 60K 65K chunk_ec_1_of_4_2_0000000000000192_00000001.liz
 68K 65K chunk_ec_1_of_4_2_00000000000001A5_00000001.liz
```
Thus if one copies files around with rsync (like @onlyjob [described above](#issuecomment-453891148) -- this sparseness must have been lost, causing space consumption).
